internal/sirius/get_firm_details_test.go

```diff
@@ -0,0 +1,122 @@
+package sirius
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/ministryofjustice/opg-sirius-supervision-firm-deputy-hub/internal/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFirmDetailsReturned(t *testing.T) {
+	mockClient := &mocks.MockClient{}
+	client, _ := NewClient(mockClient, "http://localhost:3000")
+
+	json := `	{
+		"id": 2,
+		"deputies": [
+	[]
+	],
+	"firmName": "Good Firm Inc",
+	"addressLine1": "10 St Hope Street",
+	"addressLine2": "Wellington",
+	"addressLine3": "",
+	"town": "London",
+	"county": "Buckinghamshire",
+	"postcode": "BU1 1TF",
+	"phoneNumber": "123123123",
+	"email": "good@firm.com",
+	"firmNumber": 100005
+	}`
+
+	r := ioutil.NopCloser(bytes.NewReader([]byte(json)))
+
+	mocks.GetDoFunc = func(*http.Request) (*http.Response, error) {
+		return &http.Response{
+			StatusCode: 200,
+			Body:       r,
+		}, nil
+	}
+
+	expectedResponse := FirmDetails{
+		ID: 2,
+		FirmName: "Good Firm Inc",
+		FirmNumber: 100005,
+		Email: "good@firm.com",
+		PhoneNumber: "123123123",
+		AddressLine1: "10 St Hope Street",
+		AddressLine2: "Wellington",
+		AddressLine3: "",
+		Town: "London",
+		County: "Buckinghamshire",
+		Postcode: "BU1 1TF",
+	}
+
+	firmDetails, err := client.GetFirmDetails(getContext(nil), 2)
+
+	assert.Equal(t, expectedResponse, firmDetails)
+	assert.Equal(t, nil, err)
+}
+
+func TestGetFirmReturnsNewStatusError(t *testing.T) {
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+	}))
+	defer svr.Close()
+
+	client, _ := NewClient(http.DefaultClient, svr.URL)
+
+	firmDetails, err := client.GetFirmDetails(getContext(nil), 1)
+
+	expectedResponse := FirmDetails{
+		ID: 0,
+		FirmName: "",
+		FirmNumber: 0,
+		Email: "",
+		PhoneNumber: "",
+		AddressLine1: "",
+		AddressLine2: "",
+		AddressLine3: "",
+		Town: "",
+		County: "",
+		Postcode: "",
+	}
```

Comment:
I expect this is just following the same pattern as the others but why are we returning a defaulted FirmDetails object on error?

Reply:
I believe because we tell go this file will return (FirmDetails, error) we always have to return those. Even if there's no error we have to return firm details and a nil for the error. If there is an error then we have to return an empty FirmDetails item. It just throws errors if you don't. I did check user management and it's what they've done throughout.

Reply:
You should return `nil` if the variable has no value. That's what you are doing in this in the event of a success anyway - `err = json.NewDecoder(resp.Body).Decode(&v)` will assign `nil` to `err` if it is decoded successfully.

Reply:
I can return an empty firmdetails object which will have a nil value, is that ok or not what you mean?

![Screenshot 2021-12-01 at 14 06 54](https://user-images.githubusercontent.com/52196833/144248755-86782431-115d-44bf-8f27-6973d77684ed.png)

Reply:
A nil value is just writing `nil` instead of supplying a value. See `GetDeputyClients` in PA Hub.